hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceManager.java

```diff
@@ -320,53 +336,89 @@ public void removePeer(String peerId) {
   }
 
+
   /**
-   * Add a normal source for the given peer on this region server. Meanwhile, add new replication
-   * queue to storage. For the newly added peer, we only need to enqueue the latest log of each wal
-   * group and do replication
+   * Add a replication source for the given peer on this region server. Peer should have been
+   * added to {@link #replicationPeers} before calling this method because we expect to find
+   * the peerIds {@link ReplicationPeer} in {@link #replicationPeers}.
    * @param peerId the id of the replication peer
-   * @return the source that was created
    */
   @VisibleForTesting
   ReplicationSourceInterface addSource(String peerId) throws IOException {
-    ReplicationPeer peer = replicationPeers.getPeer(peerId);
-    ReplicationSourceInterface src = createSource(peerId, peer);
-    // synchronized on latestPaths to avoid missing the new log
+    return addSource(replicationPeers.getPeer(peerId));
+  }
+
+  /**
+   * Adds a replication source. Creates source if not present otherwise fetches current source
+   * from {@link #sources}.
+   * @see #addSource(String) for method to add peer AND persist to replication store.
+   */
+  private ReplicationSourceInterface addSource(ReplicationPeer peer) throws IOException {
+    String id = peer.getId();
+    ReplicationSourceInterface src;
+    // Synchronized on latestPaths to avoid missing the new WAL.
     synchronized (this.latestPaths) {
-      this.sources.put(peerId, src);
+      src = this.sources.get(id);
+      if (src == null) {
+        // The created source may not be added if we are beat to the creation by another. It is
+        // initialized only so it should be safe to let it go; nothing to cleanup if not started.
+        src = createSource(id, peer);
+        ReplicationSourceInterface newSrc = this.sources.putIfAbsent(id, src);
+        if (newSrc != null) {
+          src.terminate("Another beat us to the installation of an instance of " + id);
+          src = newSrc;
+        }
+      }
       Map<String, NavigableSet<String>> walsByGroup = new HashMap<>();
-      this.walsById.put(peerId, walsByGroup);
+      this.walsById.put(id, walsByGroup);
       // Add the latest wal to that source's queue
-      if (this.latestPaths.size() > 0) {
-        for (Path logPath : latestPaths) {
+      if (!this.latestPaths.isEmpty()) {
+        for (Path logPath: this.latestPaths) {
           String name = logPath.getName();
-          String walPrefix = AbstractFSWALProvider.getWALPrefixFromWALName(name);
           NavigableSet<String> logs = new TreeSet<>();
           logs.add(name);
-          walsByGroup.put(walPrefix, logs);
-          // Abort RS and throw exception to make add peer failed
-          abortAndThrowIOExceptionWhenFail(
-            () -> this.queueStorage.addWAL(server.getServerName(), peerId, name));
+          walsByGroup.put(AbstractFSWALProvider.getWALPrefixFromWALName(name), logs);
+          if (src.isQueueStored()) {
```

Comment:
I wonder whether we really need this flag here, just set queueStoreage to null, or have an Optional<ReplicationQueueStorage>?

Reply:
queueStorage is from RSManager. Can't set it to null if multiple ReplicationSources and some use storage and others do not.